### PR TITLE
feat: added function to sample an array of batched indices

### PR DESCRIPTION
**Pull request by db091756**
### PR Type

- Feature
- Tests

### Description

Adds a function to sample an array of batch indices for coresubset methods.

### How Has This Been Tested?

 Additional unit tests pass.

### Does this PR introduce a breaking change?

(Write your answer here.)

### Screenshots

(Write your answer here.)

### Checklist before requesting a review
- [ ] I have made sure that my PR is not a duplicate.
- [ ] My code follows the style guidelines of this project.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have performed a self-review of my code.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

**Review by tc85324 — COMMENTED** [review]
Maybe I'm missing something, but can't we do this with:
```python
jr.choice(random_key, data_size, (num_batches, batch_size))
```

**Comment by db091756** [comment]
> Maybe I'm missing something, but can't we do this with:
> 
> ```python
> jr.choice(random_key, data_size, (num_batches, batch_size))
> ```

It's not quite the same as the current code ensures that each column (corresponding to each batch of indices for each iteration of the algorithm) contains unique indices to avoid wasted work. 

image

The above image shows what I mean. The second column of the suggested code would result in wasted work, whereas sample_batch_indices gives columns with unique elements.

File: coreax/util.py [hunk 1]
```diff
@@ -244,6 +245,41 @@ def solve_qp(kernel_mm: ArrayLike, gramian_row_mean: ArrayLike, **osqp_kwargs) -
...
+    if data_size < batch_size:
+        raise ValueError("data_size must be greater than or equal to batch_size")
+    if (data_size <= 0.0) or not isinstance(data_size, int):
+        raise ValueError("data_size must be a positive integer")
+    if (batch_size <= 0.0) or not isinstance(batch_size, int):
+        raise ValueError("batch_size must be a positive integer")
+    if (num_batches <= 0.0) or not isinstance(num_batches, int):
+        raise ValueError("num_batches must be a positive integer")
+
+    return permutation(
```
Comment (tc85324) [inline coreax/util.py:275]: I think this would be more idiomatic, and the logic easier to follow, if this were written:
```python
batch_keys = jr.split(key, num_batches)
batch_permutation = jax.vmap(jr.permutation, in_axes=(0, None))
return batch_permutation(batch_keys, data_size)
```
We don't need to set `independant=True`, because we only consider one vector at a time (inside the vmap).

File: coreax/util.py [hunk 2]
```diff
@@ -244,6 +245,41 @@ def solve_qp(kernel_mm: ArrayLike, gramian_row_mean: ArrayLike, **osqp_kwargs) -
...
+    Each column (batch) of the sampled array will contain unique elements. The largest
+    possible index sampled is dictated by data_size.
+
+    :param random_key: Key for random number generation
+    :param data_size: Size of the data we wish to sample from
+    :param batch_size: Size of the batch we wish to sample
+    :param num_batches: Number of batches to sample
+    :return: Array of batch indices of size `batch_size` x `num_batches`
+    """
+    if data_size < batch_size:
```
Comment (tc85324) [inline coreax/util.py:266]: I believe the approach we have been taking so far is to only raise custom messages when the underlying code doesn't raise anything helpful/obvious to the user, and to be as permissive as possible.

By this I mean that, for example: if `data_size=1.0`, it will not pass the isinstance check, but it would pass conversion to an int, `int(1.0)`. While the type hint restricts to integers, we can reasonably accept any value that can be cast to an integer. The caveat here is that an 'invalid' value of say '1.5' will be silently cast to the integer `1.0`, which may be unexpected. However, the type hint is already indicating that behaviour is undefined for anything that isn't an `int`, we're just trying to be as lenient as possible. This is the same reason why we use try-except in many places, over explicit type checks.

File: tests/unit/test_util.py [hunk 3]
```diff
@@ -187,6 +189,50 @@ def test_solve_qp_invalid_gramian_row_mean(self) -> None:
...
+    def test_sample_batch_indices(
+        self,
+        data_size: int,
+        batch_size: int,
+        num_batches: int,
+    ) -> None:
+        """
+        Test sample_batch_indices for valid input parameters.
+        """
+        with pytest.raises(ValueError):
```
Comment (tc85324) [inline tests/unit/test_util.py:228]: This test should also check the desired/required properties of the result (uniqueness of the indices, etc...)

**Review by tc85324 — CHANGES_REQUESTED** [review]
Thanks for your explanation, I understand the utility now! My comments are more stylistic than anything; the test probably should be stricter though.

File: tests/unit/test_util.py (hunk 3 above)
Comment (db091756) [reply tests/unit/test_util.py:212]: Added uniqueness test

File: coreax/util.py (hunk 2 above)
Comment (db091756) [reply coreax/util.py:266]: Makes sense, removed the unnecessary checks. Left the ones in that don't raise errors but behave in a weird way. Removed the corresponding tests

File: coreax/util.py (hunk 1 above)
Comment (db091756) [reply coreax/util.py]: Much nicer! Updated the code above to ensure the array comes out the right size.

File: tests/unit/test_util.py [hunk 4]
```diff
@@ -235,6 +259,22 @@ def test_sample_batch_indices(
...
+        data_size = 100
+        batch_size = 50
+        num_batches = 10
+        batch_indices = sample_batch_indices(
+            random_key=key(0),
+            data_size=data_size,
+            batch_size=batch_size,
+            num_batches=num_batches,
+        )
+        for row in batch_indices:
```
[the remark this answers is not in the document]
Comment (tc85324) [reply tests/unit/test_util.py:275]: We currently have an intra-row check (uniqueness of indices within the row), but not an inter-row uniqueness check (uniqueness between the rows). I assume this is because inter-row uniqueness is not guaranteed. I.E we would accept `[1,2],[1,2],[2,1],[1,2]` for four batches of size two, where data_size is also two. My only concern is that the test currently permits a function that generates the same order of indices for every batch.

I think the probability of sampling any specific (ordered) batch of indices is `1 / math.perm(data_size, batch_size)`; with the parameters in this test, the probability is on the order of 1e-94 (the parameters could be chosen to make this probability arbitrarily small). I believe this allows us to reasonably expect all rows to be unique (perhaps leave a comment in the test explaining why this is expected).

The test for the above could be something like `assert jnp.unique(batch_indices, axis=0).shape[0] == num_batches` . This will implicitly check the shape of `batch_indices` if you don't want a dedicate shape assertion.

**Review by tc85324 — CHANGES_REQUESTED** [review]
Apologies for taking time to get back to this. I'll be finishing existing reviews and PRs at a slower pace, but still intend to get them all completed. Thank you for responding to my previous comments. Most of my new comments are somewhat pedantic nitpicks (sorry!); the biggest issue is what I believe to be the accidental inclusion of regularised array inversion in this PR.

File: coreax/util.py [hunk 5]
```diff
@@ -245,39 +245,72 @@ def solve_qp(kernel_mm: ArrayLike, gramian_row_mean: ArrayLike, **osqp_kwargs) -
     return sol.primal
 
 
+def invert_regularised_array(
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/util.py:248]: This has been merged in now

File: tests/unit/test_util.py [hunk 6]
```diff
@@ -191,34 +192,57 @@ def test_solve_qp_invalid_gramian_row_mean(self) -> None:
...
+        """Test that `invert_regularised_array` handles invalid inputs."""
+        with pytest.raises(ValueError):
+            invert_regularised_array(
+                array=array,
+                regularisation_parameter=1e-6,
+                identity=identity,
+                rcond=rcond,
+            )
+
+    def test_invert_regularised_array(self) -> None:
```
[the remark this answers is not in the document]
Comment (db091756) [reply tests/unit/test_util.py:215]: Merged in now

File: tests/unit/test_util.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (db091756) [reply tests/unit/test_util.py:275]: Added tests that cover these cases

**Review by tc85324 — CHANGES_REQUESTED** [review]
This looks good! One minor nitpick on the function docstring then I'm happy to merge

File: coreax/util.py [hunk 7]
```diff
@@ -245,39 +245,72 @@ def solve_qp(kernel_mm: ArrayLike, gramian_row_mean: ArrayLike, **osqp_kwargs) -
...
+    )[0]
+
+
 def sample_batch_indices(
     random_key: KeyArrayLike,
     data_size: int,
     batch_size: int,
     num_batches: int,
-) -> ArrayLike:
+) -> Array:
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/util.py:292]: Good spot! Fixed

**Review by tc85324 — APPROVED** [review]
